Add clickable line numbers to console view

timbrown5 opened the pull request:
The aim was to add clickable line numbers to the console view (#43), but I hit some issues that required a refactor.

Fixes #43 

Changes:
- All API calls happen in PipelineConsole.
-- Data is passed down to children (they don't call API).
- DataTreeView no longer has state (all state is in PipelineConsole).
- Added (and switch to) function to get all steps in one go.
-- Reduces the number of API calls required.
-- Reduces the number of renders on startup.
-- Getting a list of steps no longer requires knowledge of the stages (these can be fetched asynchronously). 
- Moved state setup to the constructor
-- Reduces number of renders on startup.
-- Downside: UI might appear blank until API calls return.
- Added a new element to ConsoleLine to house line number
-- It's worth considering if we want the API to be a set of lines, instead of a wall of text. This would allow getting x lines to start with and having an element to load the next x lines.
--- x could be configurable - e.g. dropdown menu to select getting of 50, 100, 200 lines per request.
-- Lines could be fetched it batches up to a limit (e.g. 10 at a time until there are 100). 
- Added some debug logging - happy to remove it if not useful to others/slowing UI.

<!-- Please describe your pull request here. -->

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira

timbrown5 commented:
<img width="813" alt="Screenshot 2022-09-15 at 17 20 45" src="https://user-images.githubusercontent.com/4447764/190456327-e43beb97-002e-4506-8dae-49e25c446e94.png">
Looking at it now, I'm not sure if some of the formatting has regressed (maybe the new layout has the wrong classes set?).

timja commented:
What I'm seeing doesn't seem to match your screenshot:

<img width="962" alt="image" src="https://user-images.githubusercontent.com/21194782/190510535-1e520dd6-f640-4bc2-86ea-f8de51d5da22.png">

Lots of spacing between log lines, and the line numbers look bad in dark theme.

Linking to log line doesn't seem to be working for me, if I link to a log line near the bottom of the page it doesn't scroll at all

timja commented:
it seems to add an extra line number at the end?

<img width="874" alt="image" src="https://user-images.githubusercontent.com/21194782/190510872-e3bca221-d616-433c-a347-b8e2c4824978.png">

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:
```diff
@@ -76,23 +45,37 @@ const getTreeItemsFromStepList = (stepsItems: StepInfo[]) => {
-    if (steps) {
-      let stepsItems = getTreeItemsFromStepList(steps);
+    var i = stepsCopy.length
+    while (i--) {
+      let step = stepsCopy[i];
+      if (step.stageId == String(stageItemData.id)) {
+        stageSteps.push(step);
+        // Remove step from local copy - can only have one parent.
+        // This should reduce the total number of loops required.
+        stepsCopy.splice(i, 1);
+      }
```
not sure if this is related but it would be good if nested parents worked right,

e.g. parallel(linux-11) -> stage(checkout(linux-11))

for an example of where it doesn't work see:
https://ci.jenkins.io/job/Plugins/job/slack-plugin/job/master/325/pipeline-console/?selected-node=25

timbrown5 replied:
I'm not sure I follow this.
This should only be affecting steps, and (I think) steps should only have one parent.

It can be removed - it's an optimisation, not a functional component.

timja replied:
(this was broken before, wasn't sure if this code was related, I can raise a separate issue for it anyway)

timja replied:
raised https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/92

timbrown5 commented:
I will have a look - I think I haven't rebased to the correct branch locally.

Also, very happy to tweak the CSS, but it's not something I'm overly familiar with 😅.

timja commented:
> I will have a look - I think I haven't rebased to the correct branch locally.
> 
> Also, very happy to tweak the CSS, but it's not something I'm overly familiar with 😅.

once it's less broken I can fiddle with CSS if needed no worries

timbrown5 commented:
Ahh, seems it's a change in Jenkins versions that causes the difference. updating to 2.346.1 I see this:
Looks like I might need to change the structure of `ConsoleLine` (adding a new `p` for each line is breaking the newer formatting.

timja commented:
Should probably use spans, see https://weekly.ci.jenkins.io/job/Demo/7/console

timbrown5 commented:
I have updated it to work with newer Jenkins and dark mode. I have also added an offset to anchors so they don't disappear behind the breadcrumbs.

There is still an issue that anchors do not work when opting for the URL (if you reload the page they seem to). I think this is that the anchor gets evaluated before the item is rendered - but not sure how to fix it.

Jenkins 2.346.1
![Screenshot 2022-09-16 at 10 00 16](https://user-images.githubusercontent.com/4447764/190600091-089061df-b0bf-475e-a135-241c1d07957b.png)

Jenkins 2.332.4 Dark
![Screenshot 2022-09-16 at 10 24 07](https://user-images.githubusercontent.com/4447764/190605132-ceb922fa-579e-4e4b-9dd3-f8fb0b01d869.png)

Jenkins 2.332.4 Default
![Screenshot 2022-09-16 at 10 25 11](https://user-images.githubusercontent.com/4447764/190605142-c41986a3-01ae-4569-b015-e1d63d77a0eb.png)

timja commented:
lint is failing (`mvn spotless:apply`)

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:
```diff
@@ -42,12 +42,8 @@ public String getUrlName() {
-  // React component state. Then I can cross-reference this when adding the child nodes instead of
-  // making a fetch call - which seems impossible to wait for the result of (maybe fetch is the
-  // wrong thing to use?)
+  // Legacy - leave incase we want to update a sub section of steps (e.g. if a stage if still
```
```suggestion
  // Legacy - leave in case we want to update a sub section of steps (e.g. if a stage is still
```

timja commented:
spotbugs failure

mikedld commented on src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:
```diff
@@ -27,47 +21,21 @@ export interface StepInfo {
-      borderLeft: `1px dashed ${fade(theme.palette.text.primary, 0.4)}`,
-    },
-    label: {
-      fontStyle: (props: StageTreeItemProps) =>
```
Maybe I don't understand something but why was this removed? Okay if you don't want synthetic stage labels italicised, but if so they shouldn't probably be italicised in the graph view either...

timbrown5 replied:
Ah, sorry. I removed the withStyle blocks because I though it was all handled in the scss.

timbrown5 commented:
Nice catch though.
I'm away for a week now, so I will have to revisit this when I'm back.

On Fri, 16 Sep 2022, 16:43 Mike Gelfand, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx
> <https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/90#discussion_r973161168>
> :
>
> >
>  // Tree Item for stages
> -const StageTreeItem = withStyles((theme: Theme) =>
> -  createStyles({
> -    iconContainer: {
> -      "& .close": {
> -        opacity: 0.3,
> -      },
> -    },
> -    // TODO: Make line show status of block (green = passed, red = failed, etc.)
> -    group: {
> -      marginLeft: 7,
> -      paddingLeft: 18,
> -      borderLeft: `1px dashed ${fade(theme.palette.text.primary, 0.4)}`,
> -    },
> -    label: {
> -      fontStyle: (props: StageTreeItemProps) =>
>
> Maybe I don't understand something but why was this removed? Okay if you
> don't want synthetic stage labels italicised, but if so they shouldn't
> probably be italicised in the graph view either...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/90#pullrequestreview-1110888861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB54FC7F6K5SBYYP6XHFY3V6SITFANCNFSM6AAAAAAQNRSROQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

timja commented:
Issues so far:

1. Line numbers with wrapping look a bit weird
<img width="829" alt="image" src="https://user-images.githubusercontent.com/21194782/190725155-aaa29c3a-34d7-499d-a73a-a7e276f87550.png">
2. I think a bit of padding between line numbers and content would be good
3. The page lost some width between the main branch, it was made quite wide to make more room for logs, intentional?
4. Linking to line number is a bit hit or miss, doesn't work on the pipeline from https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/92 (not a blocker)
5. Extra line at the end which looks a bit weird (blue ocean doesn't have this extra line)
6. CI build is failing

timbrown5 commented:
@timja thanks for the list. I feel I should have caught some of these things 😓.
I'm posting some links to remind help me when I get back (as I'm sure I will have forgotten everything by then 😅).
https://stackoverflow.com/questions/40842277/create-line-numbers-on-pre-with-css-only#40842325 - counts as CSS (can they be clickable anchors? Maybe with offset to avoid needing component to be loaded).
This might fix a few of the issues above 
If not https://css-tricks.com/multi-line-padded-text/

Is 3 a change in the default width of the tree view?

timja commented:
> Is 3 a change in the default width of the tree view?
It's changed vs the main branch, I assume some CSS is missing now or something

timbrown5 commented:
>> Is 3 a change in the default width of the tree view?
> It's changed vs the main branch, I assume some CSS is missing now or something
I've had a look at the main branch locally and don't see the difference:

`jenkins-ci/main` [f8a7e8cfaa34c9eb7f5eb9235a716cd82b0a55fc]
<img width="935" alt="Screenshot 2022-09-23 at 11 27 03" src="https://user-images.githubusercontent.com/4447764/191941878-1bba096b-dbab-4b1c-93bf-4607dca49af8.png">
`console-line-numbers` [90017e2d8ecc8fd99861ef67d2ffb900cb24d307]
<img width="1040" alt="Screenshot 2022-09-23 at 11 21 07" src="https://user-images.githubusercontent.com/4447764/191941886-bd309cfc-61be-4bad-bc63-c4573f3567e7.png">

That said I have rebased, so maybe that has fixed it?
I also added more padding (should be the same amount as Blue Ocean).

<img width="1663" alt="Screenshot 2022-09-23 at 10 51 00" src="https://user-images.githubusercontent.com/4447764/191942209-046d66f7-7a9f-47f9-858f-4fae9f6a5451.png">

timja commented:
looks nice from screenshots.

would be good to get rid of that trailing line if possible

![image](https://user-images.githubusercontent.com/21194782/191945406-015bdf05-eea1-4988-8b0c-4ac713c371f7.png)

timbrown5 commented:
The only outstanding issue should be text wrapping (which I forgot to look at).

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx:
```diff
@@ -69,23 +68,37 @@ const getTreeItemsFromStepList = (stepsItems: StepInfo[]) => {
         key={stepItemData.id}
         nodeId={String(stepItemData.id)}
         label={stepItemData.name}
+        stageId={String(stepItemData.stageId)}
```
there's an error in the browser console that is likely this line (although could be another)

<img width="758" alt="image" src="https://user-images.githubusercontent.com/21194782/192383832-5f685760-2e2f-449a-b0b7-4dd238a62f38.png">

timja commented on src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:
```diff
@@ -67,12 +63,24 @@ public void getSteps(StaplerRequest req, StaplerResponse rsp) throws IOException
+    if (logger.isDebugEnabled()) {
+      logger.debug("Steps: '" + stepsJson + "'.");
+    }
+    rsp.getWriter().append(stepsJson);
```
Content-Type header should be added to the response

timbrown5 replied:
Have switched to using HttpResponses.okJSON/errorJSON.

timja reviewed (changes requested):
ANSI color highlighting is broken:

<img width="1362" alt="image" src="https://user-images.githubusercontent.com/21194782/192384332-502f8085-5b99-47f0-a854-9ff9e28a957e.png">


main branch:

<img width="1339" alt="image" src="https://user-images.githubusercontent.com/21194782/192384363-f271e0cd-c42c-43b7-bcea-e3376626dbf2.png">

Also steps are out of order, build should be first but archive is.

This is built from the multi-branch job at:
https://github.com/timja-org/junit-attachments-test/tree/master

timbrown5 commented:
The refactor also seems to fix some of the performance issues (at least locally). I think this is due to allowing the fetch calls to be asynchronous.

Here are some charts from firefox (using the pipeline [from here](https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/39#issuecomment-891075485)).

Before:
![Screenshot 2022-09-27 at 09 59 20](https://user-images.githubusercontent.com/4447764/192483539-7aa323cc-d8bd-4fa0-a01f-ba45098db1b7.png)

After:
![Screenshot 2022-09-27 at 09 59 31](https://user-images.githubusercontent.com/4447764/192483568-4eb7e334-8f03-4efd-829d-04fb133cbbc1.png)

timbrown5 commented:
Fixed ANSI colours:
![Screenshot 2022-09-27 at 13 37 21](https://user-images.githubusercontent.com/4447764/192528471-2e8df336-2255-4a75-a6f4-fc33d2789268.png)

Tested with:
```
pipeline {
    agent any
    options {
        ansiColor('xterm')
    }
    stages {
        stage('Build') {
            steps {
                echo '\033[34mHello\033[0m \033[33mcolorful\033[0m \033[35mworld!\033[0m'
            }
        }
    }
}
```

timja commented:
Currently getting in browser console:

Page load:
<img width="768" alt="image" src="https://user-images.githubusercontent.com/21194782/192535243-cb3f3ff7-f07d-44f0-9435-b445505a43e9.png">

When clicking on a step:
<img width="337" alt="image" src="https://user-images.githubusercontent.com/21194782/192535494-323c5c40-130a-4144-aa44-2b22a297792c.png">


same pipeline as before
https://github.com/timja-org/junit-attachments-test/tree/master?rgh-link-date=2022-09-26T21%3A31%3A53Z

timja commented:
All other issues seem fixed that I found, nice work!

timbrown5 commented:
> Currently getting in browser console:
> 
> Page load: <img alt="image" width="768" src="https://user-images.githubusercontent.com/21194782/192535243-cb3f3ff7-f07d-44f0-9435-b445505a43e9.png">
> 
> When clicking on a step: <img alt="image" width="337" src="https://user-images.githubusercontent.com/21194782/192535494-323c5c40-130a-4144-aa44-2b22a297792c.png">
> 
> same pipeline as before https://github.com/timja-org/junit-attachments-test/tree/master?rgh-link-date=2022-09-26T21%3A31%3A53Z

TIL: `npm run build:dev`

timja reviewed (approved):
LGTM, one non-blocking suggestion would be should clicking on a line highlight it? / hovering over the line number highlight the line?
